misk/src/test/kotlin/misk/web/WebDispatchTest.kt

```diff
@@ -0,0 +1,134 @@
+                WebModule(),
+                TestModule(),
+                MiskModule()
+        )
+
+        injector.injectMembers(this)
+        serviceManager.startAsync()
+        serviceManager.awaitHealthy(5, TimeUnit.SECONDS)
+    }
+
+    @After
+    fun after() {
+        serviceManager.stopAsync()
+        serviceManager.awaitStopped(5, TimeUnit.SECONDS)
+    }
+
+    @Test
+    fun post() {
+        val requestContent = helloByeJsonAdapter.toJson(HelloBye("my friend"))
+        val httpClient = OkHttpClient()
+        val request = Request.Builder()
+                .post(RequestBody.create(jsonMediaType, requestContent))
+                .url("${serverUrl}hello")
+                .build()
+
+        val response = httpClient.newCall(request).execute()
+        assertThat(response.code()).isEqualTo(200)
+        val responseContent  = response.body()!!.source()
+        assertThat(helloByeJsonAdapter.fromJson(responseContent)!!.message).isEqualTo("post hello my friend")
+    }
+
+    @Test
+    fun get() {
+        val httpClient = OkHttpClient()
+        val request = Request.Builder()
+                .get()
+                .url("${serverUrl}hello/my_friend")
+                .build()
+
+        val response = httpClient.newCall(request).execute()
+        assertThat(response.code()).isEqualTo(200)
+        val responseContent  = response.body()!!.source()
+        assertThat(helloByeJsonAdapter.fromJson(responseContent)!!.message).isEqualTo("get hello my_friend")
+    }
+
+    class TestModule : KAbstractModule() {
+        override fun configure() {
+            install(WebActionModule.create<PostHello>())
+            install(WebActionModule.create<GetHello>())
+            install(WebActionModule.create<PostBye>())
+            install(WebActionModule.create<GetBye>())
+        }
+
+        @Provides
+        @Singleton
+        fun provideWebConfig() = WebConfig(PortPicker.getFreePort(), 5000)
+    }
+
+    class PostHello : WebAction {
+        @Post("/hello")
+        @JsonResponseBody
+        fun postHello(@JsonRequestBody request: HelloBye) =
+                HelloBye("post hello ${request.message}")
+
+    }
+
+    class GetHello : WebAction {
+        @Get("/hello/{message}")
+        @JsonResponseBody
+        fun postHello(@PathParam("message") message: String) =
+                HelloBye("get hello $message")
+    }
+
+    class PostBye : WebAction {
+        @Post("/bye")
+        @JsonResponseBody
+        fun postBye(@JsonRequestBody request: HelloBye) =
+                HelloBye("post bye ${request.message}")
+    }
+
+    class GetBye : WebAction {
+        @Get("/bye/{message}")
+        @JsonResponseBody
+        fun getBye(@PathParam("message") message: String) =
+                HelloBye("get bye $message")
+    }
+
+    private val serverUrl: HttpUrl
+        get() = HttpUrl.Builder()
+                .scheme("http")
+                .host("127.0.0.1")
+                .port(webConfig.port)
```

Comment:
What happens if we set the port to say, `0` and just interrogate Jetty on which port it actually picked?

The way it works in MockWebServer is slightly neater than that even, since we rely on it to tell us both the port it picked and also the loopback IP address. This turns out to be relevant in weird network configs sadly.

Reply:
You can't interrogate Jetty, sadly. Well, I guess maybe you could peek into the connectors list and find the right connector, lemme see.